src/VertexArray.cc

```diff
@@ -0,0 +1,488 @@
+#include <iostream>
+#include <iomanip>
+#include <sstream>
+#include <vector>
+#include <memory>
+#include <cstdio>
+#include <functional>
+#include "enums.h"
+#include "VertexArray.h"
+
+#include "printutils.h"
+
+void addAttributeValues(IAttributeData &) {}
+
+void VertexData::fillInterleavedBuffer(GLbyte* interleaved_buffer) const
+{
+	// All attribute vectors need to be the same size to interleave
+	if (attributes_.size()) {
+		size_t idx = 0, last_size = attributes_[0]->size() / attributes_[0]->count(), stride = stride_;
+#if 0
+	for (const auto &data : attributes_) {
+		if (idx != 0) {
+			if (last_size != data->size() / data->count()) {
+				PRINTDB("attribute data for vertex incorrect size at index %d = %d", idx % (data->size() / data->count()));
+				PRINTDB("last_size = %d", last_size);
+				assert(false);
+			}
+		}
+		last_size = data->size() / data->count();
+		idx++;
+		//stride += data->sizeofAttribute();
+	}
+#endif // 0
+
+		GLbyte* dst_start = interleaved_buffer;
+		for (const auto &data : attributes_) {
+			size_t size = data->sizeofAttribute();
+			GLbyte* dst = dst_start;
+			const GLbyte* src = data->toBytes();
+			for (size_t i = 0; i < last_size; ++i) {
+				std::memcpy((void *)dst, (void *)src, size);
+				src += size;
+				dst += stride;
+			}
+			dst_start += size;
+		}
+	}
+}
+
+void VertexData::getLastVertex(std::vector<GLbyte> &interleaved_buffer) const
+{
+	// All attribute vectors need to be the same size to interleave
+#if 0	
+	size_t idx = 0, last_size = 0;
+	for (const auto &data : attributes_) {
+		if (idx != 0) {
+			if (last_size != data->size() / data->count()) {
+				PRINTDB("attribute data for vertex incorrect size at index %d = %d", idx % (data->size()/data->count()));
+				PRINTDB("last_size = %d", last_size);
+				assert(false);
+			}
+		}
+		last_size = data->size() / data->count();
+		idx++;
+	}
+#endif // 0
+
+	GLbyte* dst_start = interleaved_buffer.data();
+	for (const auto &data : attributes_) {
+		size_t size = data->sizeofAttribute();
+		GLbyte* dst = dst_start;
+		const GLbyte* src = data->toBytes() + data->sizeInBytes() - data->sizeofAttribute();
+		std::memcpy((void *)dst, (void *)src, size);
+		dst_start += size;
+	}
+}
+
+void VertexData::remove(size_t count)
+{
+	for (const auto &data : attributes_) {
+		data->remove(count);
+	}
+}
+
+void VertexData::createInterleavedVBO(GLuint &vbo) const
+{
+	size_t total_bytes = this->sizeInBytes();
+	if (total_bytes) {
+		GLbyte* interleaved_buffer = new GLbyte[total_bytes];
```

Comment:
It was originally, but this was part of speed updates related to debug mode where optimization is not on. It would need to be `resize()`, and then it can still use `memcpy` that should be faster than pushing every byte even in debug mode.